morpheus/_lib/include/morpheus/messages/control.hpp

```diff
@@ -271,43 +320,189 @@ class ControlMessage
      */
     void task_type(ControlMessageType task_type);
 
+    /**
+     * @brief Sets a timestamp for a specific key within a given group.
+     *
+     * This method stores a nanosecond precision timestamp associated with a unique
+     * identifier composed of a group and key. If the group::key already exists,
+     * its timestamp will be updated to the new value.
+     *
+     * @param key The specific key for which the timestamp is to be set.
+     * @param group The group to which the key belongs, aiding in namespace separation.
+     * @param timestamp_ns The timestamp value in nanoseconds to be associated with the group::key.
+     */
+    void set_timestamp(const std::string& group, const std::string& key, std::size_t timestamp_ns);
+
+    /**
+     * @brief Retrieves the timestamp for a specific key within a given group.
+     *
+     * Attempts to find and return the timestamp associated with the specified group and key.
+     * If the group::key combination does not exist, the method's behavior is determined by
+     * the fail_if_nonexist flag.
+     *
+     * @param group The group to which the key belongs.
+     * @param key The specific key for which the timestamp is requested.
+     * @param fail_if_nonexist If true, the method throws an exception if the timestamp doesn't exist.
+     *                         If false, returns std::nullopt for non-existing timestamps.
+     * @return An optional containing the timestamp value if found, or std::nullopt otherwise.
+     */
+    std::optional<std::size_t> get_timestamp(const std::string& group, const std::string& key, bool fail_if_nonexist);
+
+    /**
+     * @brief Retrieves timestamps for all keys within a given group that match a regex pattern.
+     *
+     * Searches within the specified group for keys that match the provided regex filter and returns
+     * a map of these keys and their associated timestamps. This allows for flexible queries within
+     * the group's namespace based on pattern matching.
+     *
+     * @param group The group within which to search for keys.
+     * @param regex_filter A regular expression pattern that keys must match to be included in the result.
+     * @return A map containing the matching group::key combinations and their timestamps. The map will
+     *         be empty if no matches are found.
+     */
+    std::map<std::string, std::size_t> get_timestamp(const std::string& group, const std::string& regex_filter);
+
   private:
     static const std::string s_config_schema;                          // NOLINT
     static std::map<std::string, ControlMessageType> s_task_type_map;  // NOLINT
 
     ControlMessageType m_cm_type{ControlMessageType::NONE};
     std::shared_ptr<MessageMeta> m_payload{nullptr};
+    std::shared_ptr<TensorMemory> m_tensors{nullptr};
 
     nlohmann::json m_tasks{};
     nlohmann::json m_config{};
+
+    std::map<std::string, std::size_t> m_timestamps{};
```

Comment:
Lets use `std::chrono` timepoints here that are backed by `size_t`. Will improve the readability without adding overhead